### ISSUE TITLE
"Refresh Token Not Found" = automatic logout after 2-3 days after login

### ISSUE THREAD

**Issue by EduardMe** [supabase/supabase-swift#486]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

About 2-3 days after logging in, the user gets automatically logged out (we are using the OTP login and use the API on iOS and macOS). Checking the error messages, we see:

`api(Auth.AuthError.APIError(msg: nil, code: nil, error: Optional("invalid_grant"), errorDescription: Optional("Invalid Refresh Token: Refresh Token Not Found"), weakPassword: nil))`

## To Reproduce

Using the Swift API of Supabase, login, use the app for 2-3 days and you will get logged out.

## Expected behavior

User should stay logged in indefinitely or as per the settings (see additional information).

## System information

- OS: macOS, iOS
- Version of supabase-swift: 2.15.2 

## Additional context

We have tried to increase the rate limits:
Token Refreshes = 184
Token Verifications = 100
Sign ups and sign ins = 50
Time-box user sessions = 0 
Inactivity timeout = 0
Access Tokens expiry time = 10800 (3 hours)
Refresh token reuse interval = 30

Let me know if the settings can cause this.

**Comment by hf** [comment]
This is likely due to a bug in your application that refreshes the token in parallel. Please see this doc for more information about such issues: https://supabase.com/docs/guides/auth/sessions#what-is-refresh-token-reuse-detection-and-what-does-it-protect-from

**Comment by hf** [comment]
Also be aware that if the user signed out from another device or platform (web) they are signed out from everywhere. You can control this with a parameter.

**Comment by EduardMe** [comment]
Ok, after about 2-3 days of testing with the changes, I'm still getting logged out automatically on macOS (2x in the last 2-3 days) with the error message:

```
api(Auth.AuthError.APIError(msg: nil, code: nil, error: Optional("invalid_grant"), errorDescription: Optional("Invalid Refresh Token: Already Used"), weakPassword: nil))
```

I haven't hit the logout button anywhere during testing. 

Any ideas how to fix it or if it's a bug in the library?

**Comment by grdsdev** [comment]
Hi @EduardMe 

do you have any extension also using Supabase? Widgets...

**Comment by EduardMe** [comment]
Just checked, we have a Widget, but it accesses only some userDefaults. 

We have shortcuts that call the CloudKit sync, possibly Supabase, however, I'm not using them. 

Today I got logged out again on macOS. I use it less on iOS, but there I didn't get logged out since testing. So it's just on macOS since the last 3 days.

On macOS I often run into some internet connection issues. It gets disconnected for a second. Might this influence it?

**Comment by grdsdev** [comment]
Thanks @EduardMe 

I'm investigating it, and will post any updates I find on this issue.

**Comment by LucasAbijmil** [comment]
I'm having the same issue since I updated my version of Supabase in my Xcode project. I don't do anything sophisticated in my app and the token settings are the default

**Comment by grdsdev** [comment]
@LucasAbijmil which version were you using before, which you wasn't having this issue?

**Comment by LucasAbijmil** [comment]
@grdsdev Previously I was in 2.5.1, then it was when I switched to 2.13.3 that the bug seems to have been introduced

**Comment by EduardMe** [comment]
@grdsdev I have some idea, but I'm not quite sure. Is it possible that if this gets called in parallel or when the Supabase library refreshes the session in the background and we call it at the same time it creates an issue?

We are using this func in Swift to check if the user is logged in:

```swift
func session(completion: @escaping (Result<Session, Error>) -> Void) {
    Task {
	    do {
		    if let currentSession = supabase.auth.currentSession, !currentSession.isExpired {
			    completion(.success(currentSession))
			    return
		    }
		    
		    let session = try await supabase.auth.session
		    completion(.success(session))
	    } catch {
		    completion(.failure(error))
	    }
    }
}
```

**Comment by grdsdev** [comment]
@EduardMe I don't think that is causing the issue, but your check is redundant, when you call the `try await supabase.auth.session` it already validates if session is expired, if it isn't, it returns it without refreshing, just like your first `if`.

Still looking at this issue...

**Comment by dpelletier2017** [comment]
This might be the bug I'm experiencing since December, the day I implemented Sign In With Apple in my SwiftUI App.
It seems to happen randomly. When I develop the app, it can happen 3-4 times in an afternoon because I launch my apps often, but other than that yes it seems to be happening every 2-3 days. It basically logs my users out from time to time. 

Ideally, I'd like the user to just login the first time and then never again unless he stops using the app for a prolonged period, or he deletes the app and reinstalls it or upgrades to a new phone.

`api(Auth.AuthError.APIError(msg: nil, code: nil, error: Optional("invalid_grant"), errorDescription: Optional("Invalid Refresh Token: Refresh Token Not Found"), weakPassword: nil))`

I've indeed seen that happen quite a lot of times.

**Comment by EduardMe** [comment]
For us the error is usually `Invalid Refresh Token: Already Used`. That's the only bigger roadblock for us right now. The rest works great.

**Comment by EduardMe** [comment]
@grdsdev Could you find anything? We are a bit struggling with this, or is there anything we can do temporarily to reduce the logouts?

**Comment by iBenjamin** [comment]
`2024-09-10T17:37:30+0800 supabase: Response: Status code: 400 Content-Length: 83
Body: {
  "error" : "invalid_grant",
  "error_description" : "Invalid Refresh Token: Already Used"
}
`
same here

**Comment by pedrommcarrasco** [comment]
Hey 👋
Could we please bring more attention to this issue? It’s a crucial part of the core workflow for many apps, and addressing it promptly would be greatly appreciated.

**Comment by rebryk** [comment]
I receive an error with 2.14.0. Some times the app works less than a day.
`2024-09-12T19:14:44-0700 error BroadcastReducer : [n/a] Failed to create realtime channel: api(Auth.AuthError.APIError(msg: nil, code: nil, error: Optional("invalid_grant"), errorDescription: Optional("Invalid Refresh Token: Already Used"), weakPassword: nil))`

I have my actor that updates the current session that might be accessed from different places:
```
  func session(from: URL? = nil) async throws -> Session {
    if let task {
      return try await task.value
    }

    task = Task {
      defer { self.task = nil }

      if let from {
        return try await supabase.auth.session(from: from)
      } else {
        return try await supabase.auth.session
      }
    }

    return try await task!.value
  }
```

**Comment by rebryk** [comment]
@EduardMe Have you managed to codesign the app with SDK 2.14+? Have you granted `Keychain Sharing` capabilities to the app?

**Comment by rebryk** [comment]
SDK v2.8.0 has same the problem

```
2024-09-13T17:49:05-0700 debug BroadcastReducer : [n/a] Channel status changed: unsubscribing
2024-09-13T18:07:28-0700 warning BackendService : [n/a] Attempt 1 failed: api(Auth.AuthError.APIError(msg: nil, code: nil, error: Optional("invalid_grant"), errorDescription: Optional("Invalid Refresh Token: Refresh Token Not Found"), weakPassword: nil))
```

@grdsdev i used v2.0 before, if i am not mistaken, and it was fine. however, not sure that it has realtime. i need realtime that is why i decided to upgrade at all

**Comment by dpelletier2017** [comment]
AFAIK the bug's been there since at least December 2023 because this is when I started designing my app and it's been a day 1 bug for me. 

Yes, this requires more attention, it is currently my only bug with Supabase, but it's a big one. It's really annoying my users. 

I only implemented Sign In With Apple FYI, in Swift/SwiftUI.

Also, logging in with a device seems to log out other devices. I always have to login between the Simulator and my real iPhone.

**Comment by dshukertjr** [comment]
These types of errors are hard to track down the cause. If anyone is able to reliably reproduce the issue, sharing the steps would greatly help us track down what might be the cause.

**Comment by EduardMe** [comment]
> @EduardMe Have you managed to codesign the app with SDK 2.14+? Have you granted `Keychain Sharing` capabilities to the app?

@rebryk Yes, I have granted Keychain Sharing, but I did this long before, didn't even know I needed to do this. Haven't done anything else to set things up. Just had to bump up the minimum version for macOS and iOS builds.

**Comment by rebryk** [comment]
@dshukertjr i can add supabase logger and wait for a few hours when it actually occurs. will it help you?

**Comment by rebryk** [comment]
@dshukertjr @grdsdev I have logs. What email should I use to send them?

**Comment by dshukertjr** [comment]
@rebryk Can you just hide what you need to hide and paste it here?

**Comment by rebryk** [comment]
@dshukertjr i don't know how to do that with .pulse log

**Comment by rebryk** [comment]
@dshukertjr in my case it happens with realtime.

```
2024-09-16T04:51:13Z [verbose] [Realtime] [Realtime/WebSocketClient.send(_:):111] Sending message: {"event":"heartbeat","payload":{},"topic":"phoenix","ref":"233"}

2024-09-16T04:51:13Z [verbose] [Realtime] [Realtime/WebSocketClient.receive():88] Received message: {"ref":"233","event":"phx_reply","payload":{"status":"ok","response":{}},"topic":"phoenix"}

2024-09-16T04:51:13Z [debug] [Realtime] [Realtime/RealtimeClientV2.onMessage(_:):328] heartbeat received

2024-09-16T04:51:16Z [verbose] [Realtime] [Realtime/WebSocketClient.receive():88] Received message: {"ref":null,"event":"system","payload":{"message":"Access token has expired: [message: \"Invalid token\", claim: \"exp\", claim_val: 1726462276]","status":"error","extension":"system","channel":"b0412dd4-fffc-4058-ab0d-2531f88fd6d9"},"topic":"realtime:b0412dd4-fffc-4058-ab0d-2531f88fd6d9"}

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeClientV2.onMessage(_:):331] Received event system for channel realtime:b0412dd4-fffc-4058-ab0d-2531f88fd6d9

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeChannelV2.onMessage(_:):210] Received message without event type: RealtimeMessageV2(joinRef: nil, ref: nil, topic: "realtime:b0412dd4-fffc-4058-ab0d-2531f88fd6d9", event: "system", payload: ["channel": b0412dd4-fffc-4058-ab0d-2531f88fd6d9, "message": Access token has expired: [message: "Invalid token", claim: "exp", claim_val: 1726462276], "status": error, "extension": system])

2024-09-16T04:51:16Z [verbose] [Realtime] [Realtime/WebSocketClient.receive():88] Received message: {"ref":"1","event":"phx_close","payload":{},"topic":"realtime:b0412dd4-fffc-4058-ab0d-2531f88fd6d9"}

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeClientV2.onMessage(_:):331] Received event phx_close for channel realtime:b0412dd4-fffc-4058-ab0d-2531f88fd6d9

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeChannelV2.unsubscribe():116] unsubscribing from channel realtime:b0412dd4-fffc-4058-ab0d-2531f88fd6d9

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeChannelV2.onPresenceChange(_:):350] Removing presence callback with id: 1

2024-09-16T04:51:16Z [verbose] [Realtime] [Realtime/WebSocketClient.send(_:):111] Sending message: {"join_ref":"1","topic":"realtime:b0412dd4-fffc-4058-ab0d-2531f88fd6d9","ref":"234","payload":{},"event":"phx_leave"}

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeChannelV2.onBroadcast(event:callback:):447] Removing broadcast callback with id: 2

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeClientV2.removeChannel(_:):229] No more subscribed channel in socket

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeClientV2.disconnect():302] Closing WebSocket connection

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeChannelV2.onBroadcast(event:callback:):447] Removing broadcast callback with id: 3

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeChannelV2.onMessage(_:):322] Unsubscribed from channel realtime:b0412dd4-fffc-4058-ab0d-2531f88fd6d9

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeClientV2.removeChannel(_:):229] No more subscribed channel in socket

2024-09-16T04:51:16Z [debug] [Realtime] [Realtime/RealtimeClientV2.disconnect():302] Closing WebSocket connection

2024-09-16T04:51:16Z [verbose] [_Helpers] [_Helpers/Request.rawFetch(_:):25] Request [2773CEC7-44CE-47DB-ADC8-B71579E7660C]: POST https://<PROJECT_ID>.supabase.co/auth/v1/token?grant_type=refresh_token
Body: {
  "refresh_token" : "Cl9XpqzeHzUxSuee1Bbzdg"
}

2024-09-16T04:51:16Z [verbose] [_Helpers] [_Helpers/Request.rawFetch(_:):45] Response [2773CEC7-44CE-47DB-ADC8-B71579E7660C]: Status code: 400 Content-Length: 94
Body: {
  "error" : "invalid_grant",
  "error_description" : "Invalid Refresh Token: Refresh Token Not Found"
}
```

**Comment by AlanDuong07** [comment]
Also have been experiencing this for quite some time in my SwiftUI app. Completely breaking functionality for users, so would appreciate someone to look at this! This is probably the most broken part of this SDK that affects the most users of it, so would appreciate prioritization of it

**Comment by grdsdev** [comment]
Hi all, Thank you for all the information posted in this issue.
I have now pushed a new [PR](https://github.com/supabase/supabase-swift/pull/561) which may not fix this issue in all scenarios, but I'm certain it will reduce the surface of this bug.

I'll keep tinkering with token refresh logic to improve it even further.

I'm appreciate your patience, and let me know if you have any concerns regarding the changes in the linked PR.

Thank you.

**Comment by EduardMe** [comment]
Thanks @grdsdev, will test it now over the next couple of days.

**Comment by grdsdev** [comment]
Hi all, can you all update to latest version (https://github.com/supabase/supabase-swift/releases/tag/v2.20.2) and verify if the issue still happens?

This version has several auth improvements, let me know, and I can close this issue.

**Comment by EduardMe** [comment]
I'm using it since the last couple of days and had one logout only so far. It feels like it has reduced, but need a few more days to know for sure.

**Comment by dpelletier2017** [comment]
I am also testing 2.20.2 since it has been released.

I don't know if it's a placebo effect, but I also feel like I'm experiencing less logouts.
I will also continue testing in the upcoming days.

**Comment by dpelletier2017** [comment]
My comment would be the same as last week.
So it's still happening, but less often.

I'm on 2.20.5 now. Here's something I logged in my homemade Telemetry API this morning, which also happens to use Supabase by the way :)
Is there maybe a better log I could send?

`
Bad ID token

api(message: "Bad ID token", errorCode: Auth.ErrorCode(rawValue: "unknown"), underlyingData: 62 bytes, underlyingResponse: <NSHTTPURLResponse: 0x303ae3220> { URL: https://[REDACTED].supabase.co/auth/v1/token?grant_type=id_token } { Status Code: 400, Headers {
    "Alt-Svc" =     (
        "h3=\":443\"; ma=86400"
    );
    "Content-Length" =     (
        62
    );
    "Content-Type" =     (
        "application/json"
    );
    Date =     (
        "Mon, 28 Oct 2024 11:24:31 GMT"
    );
    Server =     (
        cloudflare
    );
    "Strict-Transport-Security" =     (
        "max-age=31536000; includeSubDomains"
    );
    Vary =     (
        "Origin, Accept-Encoding"
    );
    "cf-cache-status" =     (
        DYNAMIC
    );
    "cf-ray" =     (
        "8d9a8f5b2c843a0a-YYZ"
    );
    "sb-gateway-version" =     (
        1
    );
    "x-envoy-upstream-service-time" =     (
        67
    );
    "x-supabase-api-version" =     (
        "2024-01-01"
    );
} })`


and here's my code. The line with "try" has failed : 

`
import AuthenticationServices
import Supabase

@Observable final class AuthHelper {
    @MainActor
    func signInWithApple() async -> Bool {
       // Step 1.
      // ...

        // Step 2. Get or Generate Token
        let idToken = Settings.shared.signInWithAppleToken

        if idToken == "" {
            await refreshSession()
        }

        // Step 3. Sign In
        do {
            // Step 3A. Open Session
            let credentials = OpenIDConnectCredentials(provider: .apple, idToken: idToken)
            let session = try await CloudDatabase.client.auth.signInWithIdToken(credentials: credentials)

            await executeSignInMethods(session: session)
            
            // Step 4. Refresh API Keys for external services
            // ...
            
            return true
        } catch {
            // Step 3B. Error - Revoke Token
            DispatchQueue.main.async {
                Settings.shared.signInWithAppleToken = ""
            }
            
            id = nil
            
            TelemetryInHouseHelper.send(.error(signal: .init(error: error)))

            return false
        }
    }


    @MainActor
    func refreshSession() async {
        do {
            let session = try await CloudDatabase.client.auth.session
            // Update Session Details
        } catch {
            TelemetryInHouseHelper.send(.error(signal: .init(error: error)))
        }
    }
`

**Comment by EduardMe** [comment]
I see it also less often, today it just happened on my iPhone that I got logged out.

**Comment by grdsdev** [comment]
@dpelletier2017 based on the code snippet, you provided, it may be possible to send an empty id token on the request, if that happens a I think the `Bad ID Token` is returned.

**Comment by EduardMe** [comment]
> @EduardMe
> 
> > I see it also less often, today it just happened on my iPhone that I got logged out.
> 
> With which error have you been logged out?

Sorry, it was running through TestFlight and I wasn't logging anything. Have enabled logging for the next time.

**Comment by dpelletier2017** [comment]
> @dpelletier2017 based on the code snippet, you provided, it may be possible to send an empty id token on the request, if that happens I think the Bad ID Token is returned.

Hmmm interesting. I will log it in my Telemetry to see if it's the case.
But I'm a bit at a loss. When the session refreshes through my **CloudDatabase.client.auth.session**, what is the new idToken with Sign In With Apple?

I've tried all four and they log me out right away 100% of the time. 

`
    // MARK: - Refresh Session
    @MainActor
    func refreshSession() async {
        do {
            let session = try await CloudDatabase.client.auth.session
            updateSessionDetails(session: session)
        } catch {
            TelemetryInHouseHelper.send(.error(signal: .init(error: error)))
        }
    }
    
    func updateSessionDetails(session: Session) {
        id = session.user.id

        /*
        Settings.shared.signInWithAppleToken = session.accessToken // This logs me out all the time
        
        // Which one is right ?
        dump("[Supabase Auth] Current Sign In With Apple Token: \(Settings.shared.signInWithAppleToken)")
        dump("[Supabase Auth] Access Token: \(session.accessToken)")
        dump("[Supabase Auth] Refresh Token: \(session.refreshToken)")
        dump("[Supabase Auth] Provider Token: \(String(describing: session.providerToken))")
        dump("[Supabase Auth] Provider Refresh Token: \(String(describing: session.providerRefreshToken))")
        */
    }
`

**Comment by grdsdev** [comment]
@dpelletier2017 I'm not sure if I followed.

Once you sign in with Apple using the ID token, the session will use the refresh token to refresh it; you don't need to update/send any ID token after that.

Could you open another issue for that if you still need help? I think this is unrelated to the initial discussion.

**Comment by grdsdev** [comment]
Since the initial issue has been drastically decreased, I'll close it.
